### PR TITLE
Version 0.4-BETA

### PR DESCRIPTION
**Pull request by TureBentzin**
#### this version brings some new additions and enhances Ingwers abilitys

Important new additions:

- updated Permissions and Messages
- "Origin" Messages
- Revamp Log4J Filter
- Vault
- NodeSystem

### CODE REVIEW

**Comment by TureBentzin** [comment]
Somethings may not be implemented yet

File: src/main/java/de/bentzin/ingwer/command/internalcommands/OpCommand.java [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+
+            });
+        } else {
+            identityPlayerCommand(commandSender, senderType, cmd, (identity, player) -> {
+
+                if (Ingwer.getStorage().containsIdentityWithUUID(player.getUniqueId().toString())) {
+                    MessageBuilder.prefixed().add(C.E, "You cant change the op-status of this player!").build().send(identity);
+                }
+
+                if (player.isOp()) {
```
Comment (Bommels05) [inline src/main/java/de/bentzin/ingwer/command/internalcommands/OpCommand.java:46]: Looks like here is missing some code?

File: src/main/java/de/bentzin/ingwer/command/node/AbstractNode.java [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+     * @implNote the CommandNode at the top of a trace always returns "root"
+     */
+    @Override
+    public final @NotNull String getName() {
+        return name;
+    }
+
+    @Override
+    public final Node<T> append(Node node) throws IllegalArgumentException {
+        Node.checkCommandNodeAndThrow(node);
```
Comment (Bommels05) [inline src/main/java/de/bentzin/ingwer/command/node/AbstractNode.java:50]: Did you don´t read your own implNote?

File: src/main/java/de/bentzin/ingwer/command/node/preset/NumberNode.java [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+        }
+    }
+
+    /**
+     * @return a collection of all values the Node should react to. This may only contain one value!
+     * @implNote values needs to contain at least one String
+     */
+    @Override
+    public Collection<String> values() {
+        return List.of("0");
```
Comment (Bommels05) [inline src/main/java/de/bentzin/ingwer/command/node/preset/NumberNode.java:69]:
```suggestion
        return List.of("3");
```

File: src/main/java/de/bentzin/ingwer/command/node/preset/UsageNodeExecutor.java [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+            AtomicBoolean first = new AtomicBoolean(true);
+            Objects.requireNonNull(node.getNodes()).
+                    forEach(first.get() ? o -> {stringBuilder.append(o);
+                        first.set(false);
+                    } : o -> stringBuilder.append(",").append(o));
+            stringBuilder.append("] ");
+        });
+         */
+        //TODO Usage
+        return new UsageNodeExecutor(MessageBuilder.prefixed().add(C.E, "Please care about the usage!").build());
```
Comment (Bommels05) [inline src/main/java/de/bentzin/ingwer/command/node/preset/UsageNodeExecutor.java:44]: `"Use ..."`

File: src/main/java/de/bentzin/ingwer/features/integrated/vault/PermissionCommand.java [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+     * perm user/group [user/group] add [permission] - add [permission] to [user/group]
+     * perm user/group [user/group] remove [permission] - remove [permission] to [user/group]
+     *
+     * perm user [user] join [group]
+     * perm user [user] leave [group]
+     */
+
+    @Override
+    public void execute(IngwerCommandSender commandSender, String[] cmd, CommandTarget senderType) {
+
```
Comment (Bommels05) [inline src/main/java/de/bentzin/ingwer/features/integrated/vault/PermissionCommand.java:25]: Looks like it is not yet implemented

File: src/main/java/de/bentzin/ingwer/features/integrated/vault/VaultFeature.java [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+    @Override
+    public boolean onLoad() {
+        getLogger().info("checking for Vault...");
+        if (isVault()) {
+            Plugin vault = getVault();
+            getLogger().info("found: " + vault.getName() + " v. " + vault.getDescription().getVersion());
+            getVault().getLogger().info("Ingwer was injected successfully! (; ");
+            return true;
+        } else {
+            getLogger().warning("vault could no be found!");
```
Comment (Bommels05) [inline src/main/java/de/bentzin/ingwer/features/integrated/vault/VaultFeature.java:53]:
```suggestion
            getLogger().warning("Vault could no be found!");
```

File: src/main/java/de/bentzin/ingwer/features/integrated/vault/VaultFeature.java [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+    private boolean setupPermissions() {
+        if (Ingwer.javaPlugin.getServer().getPluginManager().getPlugin("Vault") == null) {
+            return false;
+        }
+        RegisteredServiceProvider<Permission> rsp = Ingwer.javaPlugin.getServer().getServicesManager().getRegistration(Permission.class);
+        perms = rsp.getProvider();
+        return perms != null;
+    }
+
+    public boolean isVault() {
```
Comment (Bommels05) [inline src/main/java/de/bentzin/ingwer/features/integrated/vault/VaultFeature.java:88]:
```suggestion
    public boolean hasVault() {
```

**Review by Bommels05 — COMMENTED** [review]
This looks like some good progress

**Pull request merged**